### PR TITLE
Decay ek as property

### PR DESCRIPTION
**Pull request by McBs**
Add initialization of ek (energy spectrum 'ek') depending on wavenumber 'k' as property

### CODE REVIEW

**Comment by McBs** [comment]
Update for DecayTurb:
Specification of the dimension within the classname is no longer necessary.
The dimension depends now on the stencil's dimension.

File: lettuce/flows/decayingturbulence.py [hunk 1]
```diff
@@ -18,254 +17,100 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
...
-        kx = np.fft.fftfreq(self.resolution, d=1 / self.resolution)
-        ky = np.fft.fftfreq(self.resolution, d=1 / self.resolution)
-        kx, ky = np.meshgrid(kx, ky)
-        kk = np.sqrt(kx ** 2 + ky ** 2)
-        kk[0][0] = 1e-16
-
-        ### Generate spectrum
-        ek = (kk) ** 4 * np.exp(-2 * (kk / self.k0) ** 2)
-        ek[0][0] = 0
+    def __generate_wavenumbers(self):
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:26]: single underscore

File: lettuce/flows/decayingturbulence.py [hunk 2]
```diff
@@ -18,254 +17,100 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
...
+        self.dimensions = self.grid[0].shape
+        frequencies = [np.fft.fftfreq(dim, d=1 / dim) for dim in self.dimensions]
+        wavenumber = np.meshgrid(*frequencies)
+        wavenorms = np.linalg.norm(wavenumber, axis=0)
+        # self.norm = self.dimensions[0] * np.sqrt(2 * np.pi) / self.dx ** 2 if self.lattice.D == 3 else self.dimensions[
+        #                                                                                                    0] / self.dx
+        self.wavenumbers = np.arange(int(np.max(wavenorms)))
+        wavemask = (wavenorms[..., None] > self.wavenumbers - 0.5) & (wavenorms[..., None] <= self.wavenumbers + 0.5)
+        return wavenorms, wavenumber, wavemask
+    def __generate_spectrum(self):
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:36]: single underscore

File: lettuce/flows/decayingturbulence.py [hunk 3]
```diff
@@ -18,254 +17,100 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
             characteristic_length_lu=resolution, characteristic_length_pu=2*np.pi,
             characteristic_velocity_pu=None
         )
+        self.wavenumber = []
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:20]: this should probably be `self.wavenumbers`

File: lettuce/flows/decayingturbulence.py [hunk 4]
```diff
@@ -18,254 +17,100 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
...
-        u1 = np.fft.fftn(u1, axes=(0, 1))
-        # real parts
-        u0_real = u0.real
-        u1_real = u1.real
-        # imaginary parts
-        u0_imag = u0.imag
-        u1_imag = u1.imag
+    def _generate_initial_velocity(self, ek, wavenumber):
+        dx = self.units.convert_length_to_pu(1.0)
+        u = np.random.random(np.array(wavenumber).shape)*2*np.pi + 0j ## schönere Dimension?
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:47]: please delete comment. looks good to me

File: lettuce/flows/decayingturbulence.py [hunk 5]
```diff
@@ -18,254 +17,100 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
...
-        u0_real_h = np.sqrt(2 / self.units.lattice.D * ek / (u0_imag ** 2 + u0_real ** 2 + 1.e-15)) * u0_real
-        u0_imag_h = np.sqrt(2 / self.units.lattice.D * ek / (u0_imag ** 2 + u0_real ** 2 + 1.e-15)) * u0_imag
-        u1_real_h = np.sqrt(2 / self.units.lattice.D * ek / (u1_imag ** 2 + u1_real ** 2 + 1.e-15)) * u1_real
-        u1_imag_h = np.sqrt(2 / self.units.lattice.D * ek / (u1_imag ** 2 + u1_real ** 2 + 1.e-15)) * u1_imag
-        u0_real_h[0][0] = 0
-        u1_real_h[0][0] = 0
-        u0_imag_h[0][0] = 0
-        u1_imag_h[0][0] = 0
+        u_real_h = [np.sqrt(2 / self.units.lattice.D * ek / (u_imag[dim] ** 2 + u_real[dim] ** 2 + 1.e-15)) * u_real[dim] for dim in range(self.units.lattice.D)]
+        u_imag_h = [np.sqrt(2 / self.units.lattice.D * ek / (u_imag[dim] ** 2 + u_real[dim] ** 2 + 1.e-15)) * u_imag[dim] for dim in range(self.units.lattice.D)]
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:57]: lines too long

File: lettuce/flows/decayingturbulence.py [hunk 6]
```diff
@@ -18,254 +17,100 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
...
-        u1 = u1_real + u1_imag * 1.0j
-        u1f = np.fft.ifftn(u1, axes=(0, 1)) * norm
-        u = np.append(u, u1f.real[None, ...], axis=0)
-
-        p = (u[0]*0)[None,...]
+        norm = self.resolution * dx ** (-2) * np.sqrt(self.units.characteristic_length_pu) if self.units.lattice.D == 3 else self.resolution / dx
 
-        umax = np.linalg.norm(u, axis=0).max()
-        self.units.characteristic_velocity_pu = umax
+        u = [(np.fft.ifftn(u_real[dim] + u_imag[dim] * 1.0j, axes=tuple((np.arange(self.units.lattice.D)))) * norm).real for dim in range(self.units.lattice.D)]
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:91]: line breaks like so:
```
u = [
    (np.fft ...)
    for dim ...
]
```
or (even better) name intermediates
```
fourier = np.fft.ifftn(....)
u = [ (fourier * norm).real ... ]

File: lettuce/flows/decayingturbulence.py [hunk 7]
```diff
@@ -18,254 +17,100 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
...
-            reynolds_number=reynolds_number, mach_number=mach_number,
-            characteristic_length_lu=resolution, characteristic_length_pu=2*np.pi,
-            characteristic_velocity_pu=None
-        )
+        return np.asarray(u)
 
-    def analytic_solution(self, x, t=0):
-        return
+    def _compute_initial_pressure(self):
+        return np.zeros(self.dimensions)[None,...]
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:96]: ??? yet to be done?

**Review by Olllom — COMMENTED** [review]
Marvelous. So much more compact.

File: lettuce/flows/decayingturbulence.py (hunk 3 above)
Comment (McBs) [reply lettuce/flows/decayingturbulence.py:20]: We have to differ between the wavenumber array for the spectrum of zero to max(wavenorms) and the wavenumberS-array of shape (res,res,res) (3d) / (res,res) (2D), which is just important for the generating of the velocity.

The first array (wavenumber) can be extracted for the spectrum

Comment (Olllom) [reply lettuce/flows/decayingturbulence.py:20]: but this attribute (`wavenumber`) is not used anywhere in the code. it remains an empty list for its whole life time.

on the other hand, there is an attribute `self.wavenumbers` that is not created in `__init__`

File: lettuce/flows/decayingturbulence.py [hunk 8]
```diff
@@ -18,254 +17,105 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
...
-
-        ### Generate spectrum
-        ek = (kk) ** 4 * np.exp(-2 * (kk / self.k0) ** 2)
-        ek[0][0] = 0
+    def _generate_wavenumbers(self):
+        self.dimensions = self.grid[0].shape
+        frequencies = [np.fft.fftfreq(dim, d=1 / dim) for dim in self.dimensions]
+        wavenumber = np.meshgrid(*frequencies)
+        wavenorms = np.linalg.norm(wavenumber, axis=0)
+        self.wavenumbers = np.arange(int(np.max(wavenorms)))
```
Comment (Olllom) [inline lettuce/flows/decayingturbulence.py:31]: here; every attribute should be mentioned in the constructor

**Pull request merged**